br/pkg/restore/util.go

```diff
@@ -599,3 +601,152 @@ func unQuoteName(name string) string {
 	name = strings.TrimPrefix(name, "`")
 	return strings.TrimSuffix(name, "`")
 }
+
+func PrefixStartKey(key []byte) []byte {
+	var sk = make([]byte, 0, len(key)+1)
+	sk = append(sk, 'z')
+	sk = append(sk, key...)
+	return sk
+}
+
+func PrefixEndKey(key []byte) []byte {
+	if len(key) == 0 {
+		return []byte{'z' + 1}
+	}
+	return PrefixStartKey(key)
+}
+
+func keyEq(a, b []byte) bool {
+	if len(a) != len(b) {
+		return false
+	}
+	for i := 0; i < len(a); i++ {
+		if a[i] != b[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func keyCmp(a, b []byte) int {
+	var length = 0
+	var chosen = 0
+	if len(a) < len(b) {
+		length = len(a)
+		chosen = -1
+	} else if len(a) == len(b) {
+		length = len(a)
+		chosen = 0
+	} else {
+		length = len(b)
+		chosen = 1
+	}
+	for i := 0; i < length; i++ {
+		if a[i] < b[i] {
+			return -1
+		} else if a[i] > b[i] {
+			return 1
+		}
+	}
+	return chosen
+}
+
+func keyCmpInterface(a, b interface{}) int {
+	return keyCmp(a.([]byte), b.([]byte))
+}
+
+type RecoverRegionInfo struct {
+	RegionId      uint64
+	RegionVersion uint64
+	StartKey      []byte
+	EndKey        []byte
+	TombStone     bool
+}
+
+func SortRecoverRegions(regions map[uint64][]*RecoverRegion) []*RecoverRegionInfo {
+	// last log term -> last index -> commit index
+	cmps := []func(a, b *RecoverRegion) int{
+		func(a, b *RecoverRegion) int {
+			return int(a.GetLastLogTerm() - b.GetLastLogTerm())
+		},
+		func(a, b *RecoverRegion) int {
+			return int(a.GetLastIndex() - b.GetLastIndex())
+		},
+		func(a, b *RecoverRegion) int {
+			return int(a.GetCommitIndex() - b.GetCommitIndex())
+		},
+	}
+
+	// Sort region peer by last log term -> last index -> commit index, and collect all regions' version.
+	var regionInfos = make([]*RecoverRegionInfo, 0, len(regions))
+	for regionId, peers := range regions {
+		sort.Slice(peers, func(i, j int) bool {
+			for _, cmp := range cmps {
+				if v := cmp(peers[i], peers[j]); v != 0 {
+					return v > 0
+				}
+			}
+			return false
+		})
+		v := peers[0].Version
+		sk := PrefixStartKey(peers[0].StartKey)
+		ek := PrefixEndKey(peers[0].EndKey)
+		regionInfos = append(regionInfos, &RecoverRegionInfo{
+			RegionId:      regionId,
+			RegionVersion: v,
+			StartKey:      sk,
+			EndKey:        ek,
+			TombStone:     peers[0].Tombstone,
+		})
+	}
+
+	sort.Slice(regionInfos, func(i, j int) bool { return regionInfos[i].RegionVersion > regionInfos[j].RegionVersion })
+	return regionInfos
+}
+
+func CheckConsistencyAndValidPeer(regionInfos []*RecoverRegionInfo) (map[uint64]struct{}, error) {
```

Comment:
maybe add a unit test for this function.

Reply:
fixed